File: December/18th December,2024/Leetcode(1475).java
import java.util.Arrays;
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer>st=new Stack<>();
        int n=prices.length;
        int sm[]=new int[n];
        // Arrays.reverse(prices);
        int left = 0, right = n - 1;  
        while (left < right) {  
            int temp = prices[left];  
            prices[left] = prices[right];  
            prices[right] = temp;  
            left++;  
            right--;  
        }  

        for (int i = 0; i < n; i++) {  
            if (st.empty()) {  
                sm[i] = 0;  
                st.push(prices[i]);  
                continue;  
            }  
            if (prices[i] >= st.peek()) {  
                sm[i] = st.peek();  
                st.push(prices[i]);  
                continue;  
            } else {  
                while (!st.empty() && prices[i] < st.peek()) {  
                    st.pop();  
                }  
                if (st.empty()) {  
                    sm[i] = 0;  
                    st.push(prices[i]);  
                } else {  
                    sm[i] = st.peek();  
                    st.push(prices[i]);  
                }  
            }  
        }  

        int[] ans = new int[n];  
        for (int i = 0; i < n; i++) {  
            ans[i] = prices[i] - sm[i];  
        }  

        left = 0;  
        right = n - 1;  
        while (left < right) {  
            int temp = ans[left];  
            ans[left] = ans[right];  
            ans[right] = temp;  
            left++;  
            right--;  
        }  

        return ans; 
    }
}

File: December/18th December,2024/Leetcode(1475).cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int>st;
        int n=prices.size();
        vector<int>sm;
        reverse(prices.begin(),prices.end());
        for(int i:prices){
            if(st.empty()){
                sm.push_back(0);
                st.push(i);
                continue;
            }
            if(i>=st.top()){
                sm.push_back(st.top());
                st.push(i);
                continue;
            }
            else{
                while(!st.empty() && i<st.top()){
                    st.pop();
                }
                if(st.empty()){
                    sm.push_back(0);
                    st.push(i);
                }
                else{
                    sm.push_back(st.top());
                    st.push(i);
                }
            }
            // sm.push_back(i);
        }
        // reverse(sm.begin(),sm.end());
        // reverse(prices)
        // for(int i:sm) cout<<i;
        vector<int>ans;
        for(int i=0;i<n;i++){
            ans.push_back(prices[i]-sm[i]);
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};